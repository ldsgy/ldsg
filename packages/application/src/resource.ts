import { Resource } from "@ldsg/resource";
import express, { Express } from "express";
import { ApplicationSpecificResourceSettings } from "./types";

type CreateApp = () => Promise<Express>;

interface AddRoutesParams {
  app: Express;
}

type AddRoutes = (params: AddRoutesParams) => Promise<void>;

export class ApplicationResource extends Resource<ApplicationSpecificResourceSettings> {
  addRoutes: AddRoutes = async (params) => {
    const { app } = params;

    const {
      settings: { name },
      getFilteredResources,
    } = this;

    const getFilteredResourcesRes = getFilteredResources({});

    const { resources } = getFilteredResourcesRes;

    resources.forEach((resource) => {
      if ("addRoute" in resource && typeof resource.addRoute === "function") {
        resource?.addRoute();
      }
    });

    app.get("/", (req, res) => {
      res.send(`Hello, ${name ?? "World"}!`);
    });
  };

  createApp: CreateApp = async () => {
    const app = express();

    await this.addRoutes({
      app,
    });

    return app;
  };
}
