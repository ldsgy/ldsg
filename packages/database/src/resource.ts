import { Resource, ResourceConstructorParams } from "@ldsg/resource";
import mongoose from "mongoose";
import { DatabaseSpecificResourceSettings } from "./types";

export class DatabaseResource extends Resource<DatabaseSpecificResourceSettings> {
  constructor(
    params: ResourceConstructorParams<DatabaseSpecificResourceSettings>
  ) {
    super(params);

    const { connect } = this;

    connect();
  }

  connect = () => {
    const {
      settings: { uri, connectOptions },
    } = this;

    return mongoose.connect(uri, connectOptions);
  };

  disconnect = () => {
    return mongoose.disconnect();
  };
}
