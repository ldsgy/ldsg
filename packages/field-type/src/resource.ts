import { Resource } from "@ldsg/resource";
import { FieldTypeSpecificResourceSettings } from "./types";

export class FieldTypeResource extends Resource<FieldTypeSpecificResourceSettings> {}
