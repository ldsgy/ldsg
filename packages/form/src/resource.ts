import { ObjectInfo, ObjectResource } from "@ldsg/object";
import { Resource } from "@ldsg/resource";
import { FormSpecificResourceSettings } from "./types";

interface GetFormInfoParams {
  /**
   * Platform
   * Such as mongoose\formily.
   */
  platform: string;
}

export interface FormInfo {
  /**
   * Form Name
   */
  name: string;

  /**
   * Input Object Info
   */
  inputObjectInfo: ObjectInfo;

  /**
   * Output Object Info
   */
  outputObjectInfo: ObjectInfo;
}

type GetFormInfo = (params: GetFormInfoParams) => FormInfo;

export class FormResource extends Resource<FormSpecificResourceSettings> {
  getFormInfo: GetFormInfo = (params) => {
    const { platform } = params;

    const {
      settings: { name },
      getResourcesFromSettings,
    } = this;

    const { inputObjectResource, outputObjectResource, workflowResource } =
      getResourcesFromSettings();

    const inputObjectInfo = (
      inputObjectResource as ObjectResource
    ).getObjectInfo({ platform });

    const outputObjectInfo = (
      outputObjectResource as ObjectResource
    ).getObjectInfo({ platform });

    const res: FormInfo = {
      name,
      inputObjectInfo,
      outputObjectInfo,
    };

    return res;
  };
}
