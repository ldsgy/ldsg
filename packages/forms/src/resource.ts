import { FormResource } from "@ldsg/form";
import { Resource } from "@ldsg/resource";
import { FormsSpecificResourceSettings } from "./types";

export class FormsResource extends Resource<FormsSpecificResourceSettings> {
  getSubFormResources = () => {
    const { id, getFilteredResources } = this;

    const { resources } = getFilteredResources({
      parentId: id,
    });

    return resources as FormResource[];
  };
}
