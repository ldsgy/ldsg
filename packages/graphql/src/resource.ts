import { ExtendExpressApp } from "@ldsg/application";
import { Resource } from "@ldsg/resource";
import { createYoga } from "graphql-yoga";
import { GraphqlSpecificResourceSettings } from "./types";

export class GraphqlResource extends Resource<GraphqlSpecificResourceSettings> {
  extendExpressApp: ExtendExpressApp = async (params) => {
    const { app } = params;

    const { graphqlEndpoint } = this.settings;

    const yoga = createYoga({
      graphqlEndpoint,
    });

    app.use(yoga.graphqlEndpoint, yoga);

    // const handler = getHandler();

    // app.get(path, handler);
  };
}
