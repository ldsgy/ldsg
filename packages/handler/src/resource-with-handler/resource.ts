import { Resource } from "@ldsg/resource";
import { SpecificResourceSettings } from "@ldsg/types";
import { HandlerResource } from "../resource";
import { ResourceWithHandlerSpecificResourceSettings } from "./resource-settings";

export class ResourceWithHandler<
  T extends SpecificResourceSettings = SpecificResourceSettings
> extends Resource<ResourceWithHandlerSpecificResourceSettings<T>> {
  getHandler = () => {
    const { settings, getFilteredResource } = this;

    const { handlerResourceId } = settings;

    const handlerResource = getFilteredResource({
      id: handlerResourceId,
    }) as HandlerResource | undefined;

    if (!handlerResource) {
      throw new Error("invalid handler resource");
    }

    const res = handlerResource.getHandler();

    return res;
  };
}
