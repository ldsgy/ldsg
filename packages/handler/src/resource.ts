import { Resource } from "@ldsg/resource";
import {
  DefaultParams,
  DefaultRes,
  Handler,
  HandlerResourceSettings,
} from "./types";

type GetHandler<
  Params extends DefaultParams = DefaultParams,
  Res extends DefaultRes = DefaultRes
> = () => Handler<Params, Res>;

export class HandlerResource extends Resource<HandlerResourceSettings> {
  getModuleId = (): string => {
    const { code, dependencies } = this.getSettings();

    let res: string;

    if (code) {
      res = this.getId();
    } else if (dependencies[0]) {
      res = dependencies[0].name;
    } else {
      throw new Error("invalid handler settings");
    }

    return res;
  };

  getHandler: GetHandler = () => {
    const moduleId = this.getModuleId();

    try {
      const requireResult = require(moduleId);

      if ("handler" in requireResult) {
        return requireResult.handler;
      } else {
        throw new Error("invalid handler in module");
      }
    } catch (error) {
      throw new Error("require module failed");
    }
  };
}
