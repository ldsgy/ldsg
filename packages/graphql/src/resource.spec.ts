import { ApplicationResource } from "@ldsg/application";
import request from "supertest";
import { GraphqlResource } from "./resource";

test("application", async () => {
  const applicationResource = new ApplicationResource({
    id: "application",
    kind: "APPLICATION",
    parentId: "ROOT",
    settings: {
      title: "应用",
      description: "应用根资源",
      name: "测试应用",
    },
  });

  new GraphqlResource({
    id: "graphql",
    kind: "GRAPHQL",
    parentId: "application",
    settings: {
      title: "",
      description: "",
      graphqlEndpoint: "/test-graphql",
    },
  });

  const app = await applicationResource.createExpressApp();

  const response = await request(app).get("/");

  expect(response.statusCode).toBe(200);

  expect(response.text).toBe("Hello, 测试应用!");

  const response2 = await request(app).get("/test-graphql");

  expect(response2.statusCode).toBe(200);

  expect(response2.text).toMatchSnapshot();
});
