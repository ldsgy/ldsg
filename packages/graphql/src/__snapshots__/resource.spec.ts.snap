// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application 1`] = `"{"errors":[{"message":"Must provide query string.","extensions":{"code":"BAD_REQUEST"}}]}"`;
